src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+    //                              BlockBytesSize * BufferRowLengthInBlocks * BufferImageHeightInBlocks,
+    //                              BlockBytesSize * BufferRowLengthInBlocks * BufferImageHeightInBlocks * ImageDepthInBlocks>
+    const core::vector4du32_SIMD regionBlockStridesInBytes = dstImageTexelBlockInfo.convert3DBlockStridesTo1DByteStrides(regionBlockStrides);
+    auto texelBlockDim = dstImageTexelBlockInfo.getDimension();
+
+    // ! Don't confuse imageExtent with subresourceSize, imageExtent is the extent of the main region to copy and the subresourceSize is the actual size of dstImage 
+    const auto imageExtent = core::vector3du32_SIMD(mainRegion.imageExtent.width, mainRegion.imageExtent.height, mainRegion.imageExtent.depth);
+    const auto imageOffset = core::vector3du32_SIMD(mainRegion.imageOffset.x, mainRegion.imageOffset.y, mainRegion.imageOffset.z);
+    const auto imageOffsetInBlocks = dstImageTexelBlockInfo.convertTexelsToBlocks(imageOffset);
+    const auto imageExtentInBlocks = dstImageTexelBlockInfo.convertTexelsToBlocks(imageExtent);
+    const core::vector4du32_SIMD imageExtentBlockStridesInBytes = dstImageTexelBlockInfo.convert3DBlockStridesTo1DByteStrides(imageExtentInBlocks);
+             
+    // region <-> region.imageSubresource.layerCount <-> imageExtentInBlocks.z <-> imageExtentInBlocks.y <-> imageExtentInBlocks.x
+    auto updateCurrentOffsets = [&]() -> void
+    {
+        if(currentBlockInRow >= imageExtentInBlocks.x) 
+        {
+            currentBlockInRow = 0u;
+            currentRowInSlice++;
+        }
+        if(currentRowInSlice >= imageExtentInBlocks.y)
+        {
+            assert(currentBlockInRow == 0);
+            currentRowInSlice = 0u;
+            currentSliceInLayer++;
+        }
+        if(currentSliceInLayer >= imageExtentInBlocks.z)
+        {
+            assert(currentBlockInRow == 0 && currentRowInSlice == 0);
+            currentSliceInLayer = 0u;
+            currentLayerInRegion++;
+        }
+        if(currentLayerInRegion >= mainRegion.imageSubresource.layerCount) 
+        {
+            assert(currentBlockInRow == 0 && currentRowInSlice == 0 && currentSliceInLayer == 0);
+            currentLayerInRegion = 0u; 
+            currentRegion++;
+        }
+    };
+    
+    // Swizzle for when promoted image has more channels that src channel, only used when srcImageFormat!=dstImageFormat
+    asset::ICPUImageView::SComponentMapping componentMapping;
+    componentMapping.r = asset::ICPUImageView::SComponentMapping::ES_ZERO;
+    componentMapping.g = asset::ICPUImageView::SComponentMapping::ES_ZERO;
+    componentMapping.b = asset::ICPUImageView::SComponentMapping::ES_ZERO;
+    componentMapping.a = asset::ICPUImageView::SComponentMapping::ES_ONE;
+    auto srcChannelCount = asset::getFormatChannelCount(srcImageFormat);
+    for (uint32_t c = 0u; c < srcChannelCount; c++)
+        componentMapping[c] = asset::ICPUImageView::SComponentMapping::ES_IDENTITY;
```

Comment:
can you use the templated (Constexpr) versions and make 4 different types to use?

Reply:
Yea I guess so, what's the catch?

Reply:
> Yea I guess so, what's the catch?

Nabla Image Filters use excessive amounts of template m"£"&*!ion, for a good reason.

Most "behaviour modifiers" are supplied via templates, so a new type is instantiated and conditional behaviour becomes compile-time-conditional not runtime-conditional.

This breaks less inlinings and code optimizations (constant folding, etc.).

A `StaticSwizzle` will cause the compiler simply to shuffle the source registers used for any inputs while a `DefaultSwizzle` will actually do a store + load to move the values around in memory.